### PR TITLE
initial sketch of observers with UKF as method

### PR DESCRIPTION
**Pull request by martinjrobins**
Proposal for observers in PyBOP, using the unscented kalman filter as the observer method. See #12

I've added:
- `pybop.Observer`: a base observer class that just observes and does nothing else
- `pybop.UnscentedKalmanFilterObserver`: a derived observer class that uses the UKF to observe
- `pybop.ObserverCost`: a cost class that takes an observer and uses this to generate a cost function (takes the log likelihood of the observer and negates it)

A rather large disadvantage of the UKF for the SPM model is that it is very slow! I need to profile it and see where its taking all the time. Theoretically it should be 80 times slower, but it seems more than this.

### CODE REVIEW

**Comment by BradyPlanden** [comment]
Hi @martinjrobins, this is looking great, thanks for the additions! I think it's proceeding in the right direction, shall I hold off on a review until your happy, or would like you like something at the current state?

Also, thanks for updating `BaseModel` with safety catches.

**Comment by martinjrobins** [comment]
yea, don't review yet, doesn't even have proper tests and I'm sure things will change once I've done that.

**Comment by martinjrobins** [comment]
ok, this is ready for review. I'm still sorting out some test failures on windows and python 3.8, but they are nothing to do with the observer class or the UKF.

btw. I'm not really sure what to do about the typing subscription error on python 3.8, any help appreciated!

File: pybop/observers/unscented_kalman.py
Comment (BradyPlanden) [inline pybop/observers/unscented_kalman.py:229]: I think we could pre-calculate `sigma_points-x` to avoid calculating it a multiple points. 
```
sigma_points_diff = sigma_points - x
```

File: pybop/observers/unscented_kalman.py
Comment (BradyPlanden) [inline pybop/observers/unscented_kalman.py:228]: Can this be replaced with a dot product?

```
x = np.dot(w_m, sigma_points).reshape(-1,1)
```

File: pybop/observers/unscented_kalman.py
Comment (BradyPlanden) [inline pybop/observers/unscented_kalman.py:242]: Excellent, we need to remember to update once that is merged & released.

File: pybop/observers/unscented_kalman.py
Comment (BradyPlanden) [inline pybop/observers/unscented_kalman.py:290]: I wonder if we can vectorise `self.f` and `self.h` to avoid using `np.apply_along_axis`? https://stackoverflow.com/questions/23849097/numpy-np-apply-along-axis-function-speed-up

File: pybop/observers/unscented_kalman.py
Comment (BradyPlanden) [inline pybop/observers/unscented_kalman.py:295]: `np.einsum` might help with performance here. Something like:

```
P = np.einsum('i,ijk,ijl->kl', w_c, sigma_points - x_minus, sigma_points_y - y_minus)
```

File: pybop/observers/unscented_kalman.py
Comment (BradyPlanden) [inline pybop/observers/unscented_kalman.py:299]: Did you try `linalg.solve` in-place of `linalg.lstsq`? Just wondering if there is stability issues with `linalg.solve` that requires us to use lstsq.

File: tests/unit/exponential_decay.py
Comment (BradyPlanden) [inline tests/unit/exponential_decay.py:8]:
```suggestion
    Exponential decay model with two parameters y0 and k
```

File: tests/unit/test_observers.py
Comment (BradyPlanden) [inline tests/unit/test_observers.py:11]:
```suggestion
    A class to test the observer class.
```

**Review by BradyPlanden — APPROVED** [review]
Excellent work, @martinjrobins! I've added a few comments and suggestions, take what is helpful. If you can add an example notebook and update the CHANGELOG.md before merging that would be great.

File: pybop/models/base_model.py
Comment (martinjrobins) [reply pybop/models/base_model.py:206]: I don't think it is necessarily seconds, its whatever units the model is using

File: pybop/models/base_model.py
Comment (martinjrobins) [reply pybop/models/base_model.py:210]: I just wanted to be explicit so anyone reading the code would know without having to consult the docs for `step`

File: pybop/models/base_model.py
Comment (BradyPlanden) [reply pybop/models/base_model.py:206]: Gotcha, might be worth adding to the comment for future reference

File: pybop/observers/unscented_kalman.py
Comment (martinjrobins) [reply pybop/observers/unscented_kalman.py:228]: perhaps? you are relying on `w_m` being broadcasted correctly which the `*` does atm. I can give `dot` a go and see if it works

File: pybop/observers/unscented_kalman.py
Comment (martinjrobins) [reply pybop/observers/unscented_kalman.py:299]: its an over-determined set of equations, you are basically trying to fit a gaussian to a (possibly very) non-gaussian distribution, so `solve` won't work

File: pybop/observers/unscented_kalman.py
Comment (martinjrobins) [reply pybop/observers/unscented_kalman.py:295]: I'd not used `einsum` before, looks good! I'll change this

File: pybop/observers/unscented_kalman.py
Comment (martinjrobins) [reply pybop/observers/unscented_kalman.py:290]: Yea, this would have to be done somewhere in PyBaMM, needs to be a way to solve many identical pybamm models all starting from different initial conditions.

File: pybop/observers/observer.py
Comment (BradyPlanden) [reply pybop/observers/observer.py:57]: Nvm on this one, just noticed the line below.

File: pybop/models/base_model.py
Comment (martinjrobins) [reply pybop/models/base_model.py:206]: yup, added this to the docstring

File: pybop/observers/observer.py
Comment (martinjrobins) [reply pybop/observers/observer.py:23]: I've added a comment to this effect

File: pybop/observers/unscented_kalman.py
Comment (martinjrobins) [reply pybop/observers/unscented_kalman.py:229]: done

File: pybop/observers/unscented_kalman.py
Comment (martinjrobins) [reply pybop/observers/unscented_kalman.py:291]: see https://github.com/pybamm-team/PyBaMM/issues/3713

File: pybop/observers/unscented_kalman.py
Comment (martinjrobins) [reply pybop/observers/unscented_kalman.py:296]: @BradyPlanden: I tried and failed at this :( the string `i, ijk, ijl` does not work because sigma_points - x_minus is only a 2D array, not 3D. I couldn't come up with the right string to first do the outer product, and then the sub over axes 1 + 3. Any ideas?

Comment (martinjrobins) [reply pybop/observers/unscented_kalman.py:296]: actually this seems to work with the sizes, but gives a different answer than the previous implementation!

```python
P = np.einsum("k,jk,lk -> jl ", w_c, sigma_points - x_minus, sigma_points_y - y_minus)
```

Comment (martinjrobins) [reply pybop/observers/unscented_kalman.py:296]: tests still pass though! I think I need a better test on the covariance calculation.....

Comment (martinjrobins) [reply pybop/observers/unscented_kalman.py]: actually the new einsum implementation is the correct answer :) fixed now

Comment (martinjrobins) [reply pybop/observers/unscented_kalman.py]: still needs some better testing. Perhaps after this is merged in I can implement a kalman filter as well. For a linear model the kalman filter and the ukf should give identical answers correct?

**Comment by martinjrobins** [comment]
@BradyPlanden, what are you thinking on python 3.8 support. I've used type subscripting a lot which 3.8 does not support. Should I get rid of the type subscripting or should we drop 3.8 support?

**Comment by martinjrobins** [comment]
> @BradyPlanden, what are you thinking on python 3.8 support. I've used type subscripting a lot which 3.8 does not support. Should I get rid of the type subscripting or should we drop 3.8 support?

never mind, I think we can just do `from __future__ import annotations`!

**Comment by BradyPlanden** [comment]
Excellent, it would be great to maintain python 3.8 support until PyBaMM drops it, which, for visibility is being discussed in [this](https://github.com/pybamm-team/PyBaMM/issues/3390) issue.

File: pybop/observers/unscented_kalman.py
Comment (martinjrobins) [reply pybop/observers/unscented_kalman.py:231]: nope, `dot` does not do the broadcasting :( 

```
ValueError: shapes (2,) and (4,2) not aligned: 2 (dim 0) != 4 (dim 0)
```

Comment (BradyPlanden) [reply pybop/observers/unscented_kalman.py:231]: Makes sense, thought it was worth a double-check. Thanks for investigating

**Comment by martinjrobins** [comment]
any ideas on the example script failure? I tried on my machine and it worked fine, doesn't seem to be anything to do the observer or ukf

**Comment by BradyPlanden** [comment]
I'm not sure why we are getting that error, it seems to be linked to convergence problems when using a fake cost/gradient in infeasible locations. Shouldn't be caused by the observer, and I think it's solved in #166.

**Review by NicolaCourtier — APPROVED** [review]
Thanks very much @martinjrobins! Looks good to me, please merge when you are happy.

**Comment by BradyPlanden** [comment]
@all-contributors please add @martinjrobins for code and tests

**Pull request merged**
